dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/domain/MetadataIdentifier.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (c) 2004-2022, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.tracker.domain;
+
+import java.util.Objects;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hisp.dhis.attribute.AttributeValue;
+import org.hisp.dhis.common.IdentifiableObject;
+import org.hisp.dhis.tracker.TrackerIdScheme;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * MetadataIdentifier represents an immutable idScheme aware identifier of
+ * metadata.
+ *
+ * idScheme=ATTRIBUTE uses the {@link #identifier} and {@link #attributeValue}
+ * to identify metadata while the other idSchemes only rely on the
+ * {@link #identifier} (UID, CODE, NAME).
+ */
+@Value
+@AllArgsConstructor( access = AccessLevel.PRIVATE )
+public class MetadataIdentifier
+{
+
+    /**
+     * Represents the idScheme the {@link #identifier} is in.
+     */
+    @JsonProperty
```

Comment:
A `MetadaIdentifier` is basically a `TrackerIdSchemeParam` with an identifier associated with it, is this right?
Instead of having attributeValue, we have attributeUid in it, it is just reversed logic

Just wondering if it would make sense then to have `TrackerIdSchemeParam` as a field in this class.
Does it make sense? And would it make the code easier?

Reply:
TrackerIdSchemeParam represents the query parameters for idSchemes like `idScheme, programIdScheme, ...` https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-master/new-tracker.html#request-parameters So it will either be UID, CODE, NAME and if its ATTRIBUTE it will be idScheme attribute and the attribute UID. It does not identify metadata.

MetadaIdentifier is identifying metadata using an idScheme and an actual identifier. Identifiers are the UID, code, name values or for idScheme ATTRIBUTE its the attribute UID and value.

So they are distinct. Also the TrackerIdSchemeParam is referring to the entire import. The MetadataIdentifier is referring to one single metadata instance.

Before we had a TrackerIdentifier which was mixing the two things, which I think is confusing.

I also think migrating the next fields, will highlight some patterns/shortcomings, so I can adapt the API later on.

Reply:
I agree and understand that they are distinct.
Talking in DHIS2 language, could we say that `TrackerIdentifier` is in fact a `TrackerIdSchemeParamInstance`?

To be honest, it is just me thinking out loud trying to understand how he code works.
Anyway, I agree that adding the next fields will highlight how to proceed

Reply:
no worries, any question is welcome 👍🏻 

I would not say so, but I might not have fully understood the DHIS2 language 😋 

Reason being is that TrackerIdSchemeParam is a configuration of the import. The MetadataIdentifier is an identifier of metadata (encapsulating the identifier idScheme).

You could in theory create a Set<Enrollment> with enrollment.programs using different idSchemes.